### PR TITLE
review endpoint, close #925

### LINKED ISSUE #925: Reviews (second generation)
* [x] ReviewDraft
* [x] Review
* [x] ReviewRatingStatistics
* [x] Get Review by ID
* [x] Get Review by Key
* [x] Query Reviews
* [x] Create Review
* [x] Update Review by ID
* [x] Update Review by Key
* [x] Set Author Name
* [x] Set Customer
* [x] Set Rating
* [x] Set Target
* [x] Set Text
* [x] Set Title
* [x] Transition State
* [x] Set Custom Type
* [x]  Set Custom Fields 
* [x] Delete Review by ID
* [x] delete by key
* [x] setKey
* [x] search stuff see tutorial
* [x] what happens with unreviewed products concerning review search?
* [x] review rating filter search model, not only average should be possible
* [x] review rating facet search model, not only average should be possible
* [x] review rating sort search model, not only average should be possible
* [x] review rating query products (average, count)
* NOT SUPPORTED BY API: review rating query product projections
* [x] review rating query channels
* [x] also usable as comment not bounded to any resource (maybe use custom fields)
* [x] check custom fields
* [x]  check product
* [ ] "If the role of a State is modified after the calculation of this field, the calculation is not updated."  - How to trigger it?
* [x] changes in channel
* [x] changes in product
* [x] messages: http://localhost:4004/http-api-projects-messages.html#review-messages
* [x] check via update actions if still is sth. kept from title/text/rating
* [x] expansion dsl
* [x] ratingsDistribution warning for empty keys -> null pointer exception in SDK user code, getOrDefault - or creating a object wrapper for this
* [ ] moving rated view to state without rating
* [x] https://github.com/sphereio/dev.commercetools.com/blob/CTP-66_reviews_endpoint/http-api-projects-messages.md#-reviewstatetransition-message old and new state optional?
* [x] test getReviewRatingStatistics also in product
* [ ] Javadoc

### PR DESCRIPTION
**Pull request by schleichardt**
@lauraluiz please review

### CODE REVIEW

**Comment by schleichardt** [comment]
Javadoc starting point http://sphereio.github.io/sphere-jvm-sdk/javadoc/feature/925/review-endpoint/io/sphere/sdk/reviews/Review.html

File: sphere-models/src/main/java/io/sphere/sdk/reviews/Review.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+/**
+  Reviews are used to evaluate products and channels.
+
+  <h3 id=review-approval-process>Review Approval Process</h3>
+
+  <p>If you do not need any approval process, skip this part.</p>
+
+ <p>If we have an approval process for a review to be used for a product or a channel, we model the approval process with a state machine.</p>
+
+ <p>First of all, we create the approved state. Then we create the initial to-approve state, which has a possible transition to the approved state:</p>
```
Comment (lauraluiz) [inline sphere-models/src/main/java/io/sphere/sdk/reviews/Review.java]: I would mention this is only done once per project (I understand) and therefore it does not necessarily need to be in the code.

File: sphere-models/src/main/java/io/sphere/sdk/reviews/Review.java [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+ {@include.example io.sphere.sdk.reviews.approvaldemo.CreateReviewStates}
+
+ Only states with {@link io.sphere.sdk.states.StateRole#REVIEW_INCLUDED_IN_STATISTICS} makes review ratings count in statistics.
+
+ <h4 id=creating-reviews>Creating reviews</h4>
+ <p>Now we can create a review in the initial state to-approve:</p>
+
+ {@include.example io.sphere.sdk.reviews.approvaldemo.CreateReviewToApprove}
+
+ <h4 id=query-to-approve-reviews>Query which reviews should be approved</h4>
```
Comment (lauraluiz) [inline sphere-models/src/main/java/io/sphere/sdk/reviews/Review.java]: "Querying which reviews should be approved" as we had the rest of the titles "creating, approving, displaying..."

File: sphere-models/src/it/java/io/sphere/sdk/reviews/approvaldemo/ApprovingAReview.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+import io.sphere.sdk.reviews.Review;
+import io.sphere.sdk.reviews.commands.ReviewUpdateCommand;
+import io.sphere.sdk.reviews.commands.updateactions.TransitionState;
+import io.sphere.sdk.states.State;
+
+import static io.sphere.sdk.test.ReferenceAssert.assertThat;
+
+public class ApprovingAReview {
+    public static Review approveReview(final BlockingSphereClient client, final Review reviewToApprove) {
+        final ResourceIdentifier<State> state = ResourceIdentifier.ofKey("approved");//we now the state by key
```
Comment (lauraluiz) [inline sphere-models/src/it/java/io/sphere/sdk/reviews/approvaldemo/ApprovingAReview.java]: "we *know* the state by key"

File: sphere-models/src/it/java/io/sphere/sdk/reviews/ReviewProductProjectionSearchTest.java [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+                m -> m.reviewRatingStatistics().count().byDesc(),
+                p -> assertThat(p.getReviewRatingStatistics().getCount()).isEqualTo(REVIEWS_PER_PRODUCT)
+        );
+    }
+
+    @Test
+    public void sortByLowestRating() {
+        checkSorting(
+                m -> m.reviewRatingStatistics().lowestRating().byAsc(),
+                p -> assertThat(p.getReviewRatingStatistics().getLowestRating()).isEqualTo(0)
```
Comment (lauraluiz) [inline sphere-models/src/it/java/io/sphere/sdk/reviews/ReviewProductProjectionSearchTest.java]: Document how can you get `0` with random ratings

File: sphere-models/src/it/java/io/sphere/sdk/reviews/ReviewProductProjectionSearchTest.java [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+        );
+    }
+
+    @Test
+    public void sortByHighestRating() {
+        checkSorting(
+                m -> m.reviewRatingStatistics().highestRating().byDesc(),
+                p -> assertThat(p.getReviewRatingStatistics().getHighestRating()).isEqualTo(HIGHEST_RATING)
```
Comment (lauraluiz) [inline sphere-models/src/it/java/io/sphere/sdk/reviews/ReviewProductProjectionSearchTest.java]: Document how can you get the highest with random ratings

File: sphere-models/src/main/java/io/sphere/sdk/reviews/search/ReviewRatingStatisticsFilterSearchModel.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
+package io.sphere.sdk.reviews.search;
+
+import io.sphere.sdk.search.model.RangeTermFilterSearchModel;
+import io.sphere.sdk.search.model.SearchModel;
+import io.sphere.sdk.search.model.SearchModelImpl;
+
+import javax.annotation.Nullable;
+import java.math.BigDecimal;
+
+public class ReviewRatingStatisticsFilterSearchModel<T> extends SearchModelImpl<T> {
```
Comment (lauraluiz) [inline sphere-models/src/main/java/io/sphere/sdk/reviews/search/ReviewRatingStatisticsFilterSearchModel.java]: Missing `ReviewRatingStatisticsFacetAndFilterSearchModel`

**Comment by lauraluiz** [comment]
Reviewed

**Pull request merged**